src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -7179,6 +7183,42 @@ QueryProcessingStage::Enum MergeTreeData::getQueryProcessingStage(
 }
 
 
+bool MergeTreeData::canUseParallelReplicasBasedOnPKAnalysis(
+    ContextPtr query_context,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info) const
+{
+    const auto & snapshot_data = assert_cast<const MergeTreeData::SnapshotData &>(*storage_snapshot->data);
+    const auto & parts = snapshot_data.parts;
+
+    MergeTreeDataSelectExecutor reader(*this);
+    auto result_ptr = reader.estimateNumMarksToRead(
+        parts,
+        query_info.prewhere_info,
+        storage_snapshot->getMetadataForQuery()->getColumns().getAll().getNames(),
+        storage_snapshot->metadata,
+        storage_snapshot->metadata,
+        query_info,
+        /*added_filter_nodes*/ActionDAGNodes{},
+        query_context,
+        query_context->getSettingsRef().max_threads);
+
+    if (result_ptr->error())
+        throw Exception(ErrorCodes::CANNOT_PERFORM_INDEX_ANALYSIS, "Cannot estimate the number of granules to read");
+
+    LOG_TRACE(log, "Estimated number of granules to read is {}", result_ptr->marks());
+
+    bool decision = result_ptr->marks() >= query_context->getSettingsRef().parallel_replicas_min_number_of_granules_to_enable;
```

Comment:
Apart from the idea of having a nº of replicas vs (on / off). Would it be possible to use nº of rows instead (or in addition) of granules?

I know granules makes more sense in terms of internals, but as a user it's much easier to reason about number of rows instead (and you might not even know the size of the granules). For example, "use one replica per 10M estimated rows to read" vs "use one replica per 1000 estimated granules to read".

Reply:
We would need to add some estimation, because we don't have such currently. And I guess `parallel_replicas_min_number_of_bytes_to_enable ` would be even better